File: backend/app/repositories/cached_product_repository.py
"""Redis-backed caching decorator over a `ProductRepository` (BE-08).

Satisfies `ProductRepository` structurally and delegates every read to an
inner one, so caching is invisible to the HTTP layer and is removed by
pointing `app.api.deps.get_sql_product_repository` back at the inner
repository.

Keys
    ``products:q:{stable_hash(ProductQuery)}`` for listings, and
    ``products:id:{product_id}`` for single products. The id key is shared
    by `get_product` and `get_product_detail` — a `ProductDetail` is a
    strict superset of `Product` — so the payload carries a ``kind``
    discriminator. A detail entry satisfies both methods (projected down
    for the summary); a summary entry satisfies only `get_product`, and a
    detail read treats it as a miss and overwrites it with the richer
    payload. Without the discriminator a summary hit would silently answer
    a detail read with no `long_description` and no reviews.

Invalidation
    None. Nothing writes to the catalogue through this application — it
    changes only via `python -m app.seed` or a direct database edit — so
    entries expire by TTL alone. The consequence is real and accepted: a
    reseed is invisible to the API for up to `CACHE_TTL_SECONDS`.

Absent results are not cached. Distinguishing "cached absent" from "cache
miss" needs a sentinel, and buys nothing: a lookup by unknown id is already
a single indexed `SELECT` returning no rows.

`RedisError` is caught around every cache read and write, so an outage
degrades to a miss or a skipped write rather than a 500. The client must
carry socket timeouts (see `app.redis_client.get_sync_redis_client`), or an
unreachable-but-not-refused host hangs the calling thread instead of raising.
`UnknownSortKeyError` is deliberately raised outside any `try`, so it
propagates and is never cached — an invalid sort key must fail every time,
not just once.
"""

from __future__ import annotations

import hashlib
import json
import logging
from collections.abc import Callable
from dataclasses import asdict, fields
from typing import Any

from redis import Redis
from redis.exceptions import RedisError

from app.domain.catalog import (
    Category,
    Product,
    ProductDetail,
    ProductPage,
    ProductQuery,
    Review,
    SortDirection,
    SortKey,
)
from app.domain.repositories import ProductRepository

logger = logging.getLogger(__name__)

_LIST_KEY_PREFIX = "products:q:"
_PRODUCT_KEY_PREFIX = "products:id:"

_KIND_PRODUCT = "product"
_KIND_PRODUCT_DETAIL = "product_detail"

#: Field names read back off a cached payload. Derived from the dataclasses
#: rather than restated, so adding a field to either type cannot leave this
#: module quietly dropping it on the way out of the cache.
_PRODUCT_FIELDS = tuple(f.name for f in fields(Product))
_PRODUCT_DETAIL_FIELDS = tuple(f.name for f in fields(ProductDetail))


def _enum_value(value: object) -> object:
    """Unwraps a `SortKey`/`SortDirection` to its `.value`, passing others through.

    `sort` and `direction` are *typed* as enums, but nothing at runtime stops
    a caller passing a raw string — the inner repositories defend against
    exactly that. A plain `str` has no `.value`, so it is used as-is, which
    keeps key derivation from raising ahead of the inner repository's own
    `UnknownSortKeyError`.
    """
    return value.value if isinstance(value, (SortKey, SortDirection)) else value


def _stable_hash(query: ProductQuery) -> str:
    """Hashes `query`'s field *values* in a fixed order.

    Not `hash(query)` (a frozen dataclass's hash depends on field declaration
    order, which is not a contract worth depending on) and not `repr(query)`
    (formatting is an implementation detail). Enums serialize by `.value`, so
    `SortKey.NAME` and `"name"` cannot hash differently, and the key stays
    stable across process restarts. Two queries with identical values but
    different keyword ordering therefore share a key.
    """
    values: tuple[Any, ...] = (
        query.name_contains,
        query.category_slug,
        _enum_value(query.sort),
        _enum_value(query.direction),
        query.limit,
        query.offset,
    )
    payload = json.dumps(values, separators=(",", ":"))
    return hashlib.sha256(payload.encode("utf-8")).hexdigest()


def _list_key(query: ProductQuery) -> str:
    return f"{_LIST_KEY_PREFIX}{_stable_hash(query)}"


def _product_key(product_id: str) -> str:
    return f"{_PRODUCT_KEY_PREFIX}{product_id}"


def _encode(product: Product) -> str:
    """Serializes a `Product`, tagging it by what it actually is.

    `asdict` walks nested dataclasses, so `category` and `reviews` come along
    for free. The kind is taken from the runtime type rather than the calling
    method: `get_product` is free to return a `ProductDetail` (Liskov), and
    when it does, storing it as a detail is strictly better than truncating.
    """
    payload = asdict(product)
    payload["kind"] = (
        _KIND_PRODUCT_DETAIL if isinstance(product, ProductDetail) else _KIND_PRODUCT
    )
    return json.dumps(payload)


def _to_product(data: dict[str, Any]) -> Product:
    """Rebuilds the summary shape, ignoring any detail-only fields present."""
    known = {name: data[name] for name in _PRODUCT_FIELDS}
    known["category"] = Category(**data["category"])
    return Product(**known)


def _to_product_detail(data: dict[str, Any]) -> ProductDetail:
    known = {name: data[name] for name in _PRODUCT_DETAIL_FIELDS}
    known["category"] = Category(**data["category"])
    known["reviews"] = tuple(Review(**review) for review in data["reviews"])
    return ProductDetail(**known)


def _decode_product(data: dict[str, Any]) -> Product | None:
    """A hit of either kind answers a summary read."""
    if data.get("kind") not in (_KIND_PRODUCT, _KIND_PRODUCT_DETAIL):
        return None
    return _to_product(data)


def _decode_product_detail(data: dict[str, Any]) -> ProductDetail | None:
    """Only a detail hit answers a detail read; a summary hit is a miss."""
    if data.get("kind") != _KIND_PRODUCT_DETAIL:
        return None
    return _to_product_detail(data)


def _encode_page(page: ProductPage) -> str:
    return json.dumps(
        {"items": [asdict(item) for item in page.items], "total": page.total}
    )


def _decode_page(data: dict[str, Any]) -> ProductPage:
    return ProductPage(
        items=tuple(_to_product(item) for item in data["items"]), total=data["total"]
    )


class CachedProductRepository:
    """`ProductRepository` decorator caching reads in Redis under a TTL.

    Holds no engine, session, or pool of its own: anything not usable from
    the cache is asked of `inner`. See the module docstring for key layout,
    invalidation, and failure handling.
    """

    def __init__(
        self, inner: ProductRepository, redis: Redis, ttl_seconds: int
    ) -> None:
        self._inner = inner
        self._redis = redis
        self._ttl_seconds = ttl_seconds

    # -- ProductRepository ---------------------------------------------

    def list_products(self, query: ProductQuery) -> ProductPage:
        key = _list_key(query)
        cached = self._read(key, _decode_page)
        if cached is not None:
            return cached

        # Outside any try/except: UnknownSortKeyError must propagate uncached.
        page = self._inner.list_products(query)
        self._write(key, _encode_page(page))
        return page

    def get_product(self, product_id: str) -> Product | None:
        key = _product_key(product_id)
        cached = self._read(key, _decode_product)
        if cached is not None:
            return cached

        product = self._inner.get_product(product_id)
        if product is not None:
            self._write(key, _encode(product))
        return product

    def get_product_detail(self, product_id: str) -> ProductDetail | None:
        key = _product_key(product_id)
        cached = self._read(key, _decode_product_detail)
        if cached is not None:
            return cached

        detail = self._inner.get_product_detail(product_id)
        if detail is not None:
            self._write(key, _encode(detail))
        return detail

    def list_categories(self) -> tuple[Category, ...]:
        """Delegates uncached: a handful of rows, one unbounded query."""
        return self._inner.list_categories()

    # -- Redis, where any failure is a miss ------------------------------

    def _read[T](
        self, key: str, decode: Callable[[dict[str, Any]], T | None]
    ) -> T | None:
        """Returns a decoded cache hit, or `None` for a miss of any cause.

        A miss, an outage, and a malformed or wrong-kind entry are all the
        same outcome to the caller — ask the inner repository — so they
        collapse here rather than at three call sites.
        """
        try:
            raw = self._redis.get(key)
        except RedisError:
            logger.warning(
                "Redis unavailable reading cache key=%s; falling back to source",
                key,
                exc_info=True,
            )
            return None

        if raw is None:
            return None
        try:
            return decode(json.loads(raw))
        except (KeyError, TypeError, ValueError):
            logger.warning("Discarding malformed cache entry for key=%s", key)
            return None

    def _write(self, key: str, value: str) -> None:
        try:
            self._redis.set(key, value, ex=self._ttl_seconds)
        except RedisError:
            logger.warning(
                "Redis unavailable writing cache key=%s; skipping cache write",
                key,
                exc_info=True,
            )

File: backend_js/src/repositories/cached-product.repository.ts
/**
 * Redis-backed caching decorator over a `ProductRepository`.
 *
 * Implements the same interface and delegates every read to an inner one, so
 * caching is invisible to the HTTP layer and is removed by handing the module
 * the inner repository directly.
 *
 * Keys
 *   `products:q:{stableHash(ProductQuery)}` for listings, and
 *   `products:id:{productId}` for single products. The id key is shared by
 *   `getProduct` and `getProductDetail` — a `ProductDetail` is a strict
 *   superset of a `Product` — so the payload carries a `kind` discriminator. A
 *   detail entry satisfies both methods (projected down for the summary); a
 *   summary entry satisfies only `getProduct`, and a detail read treats it as
 *   a miss and overwrites it with the richer payload. Without the
 *   discriminator a summary hit would silently answer a detail read with no
 *   `longDescription` and no reviews.
 *
 * Invalidation
 *   None. Nothing writes to the catalogue through this application — it
 *   changes only via the seeder or a direct database edit — so entries expire
 *   by TTL alone. The consequence is real and accepted: a reseed is invisible
 *   to the API for up to `CACHE_TTL_SECONDS`.
 *
 * Absent results are not cached. Distinguishing "cached absent" from "cache
 * miss" needs a sentinel, and buys nothing: a lookup by unknown id is already
 * a single indexed `SELECT` returning no rows.
 *
 * Every cache read and write is wrapped, so a Redis outage degrades to a miss
 * or a skipped write rather than a 500. `UnknownSortKeyError` is deliberately
 * thrown outside any of that, so it propagates and is never cached — an
 * invalid sort key must fail every time, not just once.
 */

import { Inject, Injectable } from '@nestjs/common'
import { createHash } from 'node:crypto'
import type { Redis } from 'ioredis'

import { JsonLogger } from '../common/logging/json-logger'
import {
  Category,
  Product,
  ProductDetail,
  ProductPage,
  ProductQuery,
  isProductDetail,
  makeReview,
} from '../domain/catalog'
import { ProductRepository } from '../domain/repositories'
import { REDIS_CLIENT } from '../redis/redis.constants'
import { settings } from '../config/settings'

const logger = new JsonLogger('app.cache')

const LIST_KEY_PREFIX = 'products:q:'
const PRODUCT_KEY_PREFIX = 'products:id:'

const KIND_PRODUCT = 'product'
const KIND_PRODUCT_DETAIL = 'product_detail'

type CachedPayload = Record<string, unknown>

/**
 * Hashes a query's field *values* in a fixed order.
 *
 * Not a hash of the object (property order is not a contract worth depending
 * on) and not its `JSON.stringify` in declaration order. Enum members
 * serialize by their string value, so the key stays stable across process
 * restarts and two queries with identical values share a key regardless of how
 * they were constructed.
 */
export function stableHash(query: ProductQuery): string {
  const values = [
    query.nameContains,
    query.categorySlug,
    query.sort,
    query.direction,
    query.limit,
    query.offset,
  ]
  return createHash('sha256').update(JSON.stringify(values), 'utf8').digest('hex')
}

function listKey(query: ProductQuery): string {
  return `${LIST_KEY_PREFIX}${stableHash(query)}`
}

function productKey(productId: string): string {
  return `${PRODUCT_KEY_PREFIX}${productId}`
}

/**
 * Serializes a product, tagging it by what it actually is.
 *
 * The kind is taken from the runtime shape rather than the calling method:
 * `getProduct` is free to return a `ProductDetail`, and when it does, storing
 * it as a detail is strictly better than truncating it.
 */
function encodeProduct(product: Product): string {
  return JSON.stringify({
    ...product,
    kind: isProductDetail(product) ? KIND_PRODUCT_DETAIL : KIND_PRODUCT,
  })
}

function toCategory(data: CachedPayload): Category {
  return { slug: String(data.slug), name: String(data.name) }
}

/** Rebuilds the summary shape, ignoring any detail-only fields present. */
function toProduct(data: CachedPayload): Product {
  return {
    id: String(data.id),
    name: String(data.name),
    priceMinor: Number(data.priceMinor),
    currency: String(data.currency),
    shortDescription: String(data.shortDescription),
    thumbnailUrl: String(data.thumbnailUrl),
    category: toCategory(data.category as CachedPayload),
  }
}

function toProductDetail(data: CachedPayload): ProductDetail {
  return {
    ...toProduct(data),
    longDescription: String(data.longDescription),
    reviews: (data.reviews as CachedPayload[]).map((review) =>
      makeReview({
        id: String(review.id),
        author: String(review.author),
        rating: Number(review.rating),
        body: String(review.body),
      }),
    ),
  }
}

/** A hit of either kind answers a summary read. */
function decodeProduct(data: CachedPayload): Product | null {
  if (data.kind !== KIND_PRODUCT && data.kind !== KIND_PRODUCT_DETAIL) {
    return null
  }
  return toProduct(data)
}

/** Only a detail hit answers a detail read; a summary hit is a miss. */
function decodeProductDetail(data: CachedPayload): ProductDetail | null {
  if (data.kind !== KIND_PRODUCT_DETAIL) {
    return null
  }
  return toProductDetail(data)
}

function encodePage(page: ProductPage): string {
  return JSON.stringify({ items: page.items, total: page.total })
}

function decodePage(data: CachedPayload): ProductPage {
  return {
    items: (data.items as CachedPayload[]).map(toProduct),
    total: Number(data.total),
  }
}

@Injectable()
export class CachedProductRepository implements ProductRepository {
  constructor(
    private readonly inner: ProductRepository,
    @Inject(REDIS_CLIENT) private readonly redis: Redis,
    private readonly ttlSeconds: number = settings.cacheTtlSeconds,
  ) {}

  // -- ProductRepository ---------------------------------------------

  async listProducts(query: ProductQuery): Promise<ProductPage> {
    const key = listKey(query)
    const cached = await this.read(key, decodePage)
    if (cached !== null) {
      return cached
    }

    // Outside the guarded read/write: UnknownSortKeyError must propagate
    // uncached.
    const page = await this.inner.listProducts(query)
    await this.write(key, encodePage(page))
    return page
  }

  async getProduct(productId: string): Promise<Product | null> {
    const key = productKey(productId)
    const cached = await this.read(key, decodeProduct)
    if (cached !== null) {
      return cached
    }

    const product = await this.inner.getProduct(productId)
    if (product !== null) {
      await this.write(key, encodeProduct(product))
    }
    return product
  }

  async getProductDetail(productId: string): Promise<ProductDetail | null> {
    const key = productKey(productId)
    const cached = await this.read(key, decodeProductDetail)
    if (cached !== null) {
      return cached
    }

    const detail = await this.inner.getProductDetail(productId)
    if (detail !== null) {
      await this.write(key, encodeProduct(detail))
    }
    return detail
  }

  /** Delegates uncached: a handful of rows, one unbounded query. */
  async listCategories(): Promise<readonly Category[]> {
    return this.inner.listCategories()
  }

  // -- Redis, where any failure is a miss ------------------------------

  /**
   * Returns a decoded cache hit, or `null` for a miss of any cause.
   *
   * A miss, an outage, and a malformed or wrong-kind entry are all the same
   * outcome to the caller — ask the inner repository — so they collapse here
   * rather than at three call sites.
   */
  private async read<T>(
    key: string,
    decode: (data: CachedPayload) => T | null,
  ): Promise<T | null> {
    let raw: string | null
    try {
      raw = await this.redis.get(key)
    } catch (cause) {
      logger.warn('redis unavailable reading cache; falling back to source', {
        key,
        error: cause instanceof Error ? cause.message : String(cause),
      })
      return null
    }

    if (raw === null) {
      return null
    }
    try {
      return decode(JSON.parse(raw) as CachedPayload)
    } catch {
      logger.warn('discarding malformed cache entry', { key })
      return null
    }
  }

  private async write(key: string, value: string): Promise<void> {
    try {
      await this.redis.set(key, value, 'EX', this.ttlSeconds)
    } catch (cause) {
      logger.warn('redis unavailable writing cache; skipping cache write', {
        key,
        error: cause instanceof Error ? cause.message : String(cause),
      })
    }
  }
}
